File: 2024/19.1.py
from functools import cache

with open("19.txt") as file:
  inputs = [line.rstrip() for line in file]

towels = []
designs = []
total = 0

parsedTowels = False
for i in inputs:
  if i == '':
    parsedTowels = True
  elif parsedTowels:
    designs.append(i)
  else:
    towels = i.split(', ')

@cache
def search(design: str):
  if design == "":
    return 1

  found = 0
  for towel in towels:
    if design.startswith(towel):
      found += search(design[len(towel):])

  return found

for design in designs:
  if search(design):
    total += 1

print(total)

File: 2024/19.1.js
const { once } = require('node:events');
const { createReadStream } = require('node:fs');
const { createInterface } = require('node:readline');

const test = { file: '19.test.txt' };
const goal = { file: '19.txt' };
const game = goal;

let towels = [];
const designs = [];
let total = 0;
const memo = {};

let parsedTowels = false;
const init = (line = '') => {
  if (line == '') {
    parsedTowels = true;
  } else if (parsedTowels) {
    designs.push(line);
  } else {
    towels = line.split(', ')
  }
}

const search = (design = '') => {
  if (design == '') return 1;

  if (memo[design]) return memo[design];

  let found = 0;
  towels.forEach(towel => {
    if (design.substring(0, towel.length) == towel) {
      found += search(design.substring(towel.length));
    }
  })

  memo[design] = found;
  return found;
}

const doWork = () => {
  designs.forEach(d => {
    console.log(d)
    const arrangements = search(d);
    if (arrangements) total++
  })
}

(async function processLineByLine() {
  try {
    const rl = createInterface({
      input: createReadStream(game.file),
      crlfDelay: Infinity,
    });

    rl.on('line', (line) => {
      init(line);
    });

    await once(rl, 'close');

    doWork()
    console.log(total)

  } catch (err) {
    console.error(err);
  }
})();
